File: src/javascript/fileDownloader.js
const axios = require('axios');
const cheerio = require('cheerio');
const fs = require('fs').promises;
const path = require('path');
const { URL } = require('url');

/**
 * Clase para descargar archivos desde páginas web
 */
class FileDownloader {
    /**
     * Constructor
     * @param {string} baseUrl - URL base del sitio web
     * @param {Object} options - Opciones de configuración
     */
    constructor(baseUrl, options = {}) {
        this.baseUrl = baseUrl;
        this.userAgent = options.userAgent || 'Mozilla/5.0 (Node.js FileDownloader)';
        this.timeout = options.timeout || 30000;
        
        // Configurar axios
        this.httpClient = axios.create({
            timeout: this.timeout,
            headers: {
                'User-Agent': this.userAgent
            }
        });
    }

    /**
     * Obtiene el contenido HTML de una página
     * @param {string} url - URL de la página
     * @returns {Promise<string>} Contenido HTML
     */
    async fetchPageContent(url) {
        try {
            const response = await this.httpClient.get(url);
            return response.data;
        } catch (error) {
            console.error(`Error obteniendo contenido de ${url}:`, error.message);
            throw error;
        }
    }

    /**
     * Extrae enlaces de archivos con una extensión específica
     * @param {string} htmlContent - Contenido HTML
     * @param {string} fileExtension - Extensión de archivo a buscar
     * @returns {Array<string>} Array de URLs de archivos
     */
    parseFileLinks(htmlContent, fileExtension) {
        const $ = cheerio.load(htmlContent);
        const fileLinks = [];

        $('a[href]').each((index, element) => {
            const href = $(element).attr('href');
            if (href && href.endsWith(fileExtension)) {
                // Convertir URL relativa a absoluta
                const absoluteUrl = this.resolveUrl(this.baseUrl, href);
                fileLinks.push(absoluteUrl);
            }
        });

        return fileLinks;
    }

    /**
     * Descarga un archivo desde una URL
     * @param {string} fileUrl - URL del archivo
     * @param {string} outputPath - Ruta donde guardar el archivo
     * @returns {Promise<boolean>} true si la descarga fue exitosa
     */
    async downloadFile(fileUrl, outputPath) {
        try {
            // Crear directorio si no existe
            const dir = path.dirname(outputPath);
            await fs.mkdir(dir, { recursive: true });

            // Descargar archivo
            const response = await this.httpClient.get(fileUrl, {
                responseType: 'stream'
            });

            // Guardar archivo
            const writer = require('fs').createWriteStream(outputPath);
            response.data.pipe(writer);

            return new Promise((resolve, reject) => {
                writer.on('finish', () => {
                    console.log(`Descargado: ${path.basename(outputPath)}`);
                    resolve(true);
                });
                writer.on('error', reject);
            });

        } catch (error) {
            console.error(`Error descargando ${fileUrl}:`, error.message);
            return false;
        }
    }

    /**
     * Descarga todos los archivos con una extensión específica desde una página
     * @param {string} pageUrl - URL de la página
     * @param {string} fileExtension - Extensión de archivos a buscar
     * @param {string} downloadDir - Directorio donde guardar los archivos
     * @returns {Promise<number>} Número de archivos descargados exitosamente
     */
    async downloadFilesFromPage(pageUrl, fileExtension, downloadDir = 'archivos_js') {
        try {
            // Obtener contenido de la página
            const htmlContent = await this.fetchPageContent(pageUrl);

            // Parsear enlaces de archivos
            const fileLinks = this.parseFileLinks(htmlContent, fileExtension);

            if (fileLinks.length === 0) {
                console.log(`No se encontraron archivos con extensión '${fileExtension}'`);
                return 0;
            }

            console.log(`Encontrados ${fileLinks.length} archivos para descargar`);

            // Descargar cada archivo
            let successfulDownloads = 0;
            for (const fileUrl of fileLinks) {
                const filename = this.extractFilename(fileUrl);
                const outputPath = path.join(downloadDir, filename);

                if (await this.downloadFile(fileUrl, outputPath)) {
                    successfulDownloads++;
                }
            }

            return successfulDownloads;

        } catch (error) {
            console.error('Error en downloadFilesFromPage:', error.message);
            return 0;
        }
    }

    /**
     * Convierte URL relativa a absoluta
     * @param {string} baseUrl - URL base
     * @param {string} relativeUrl - URL relativa
     * @returns {string} URL absoluta
     */
    resolveUrl(baseUrl, relativeUrl) {
        try {
            return new URL(relativeUrl, baseUrl).href;
        } catch (error) {
            console.error('Error resolviendo URL:', error.message);
            return relativeUrl;
        }
    }

    /**
     * Extrae el nombre del archivo de una URL
     * @param {string} url - URL del archivo
     * @returns {string} Nombre del archivo
     */
    extractFilename(url) {
        try {
            const urlObj = new URL(url);
            const pathname = urlObj.pathname;
            const filename = path.basename(pathname);
            return filename || 'downloaded_file';
        } catch (error) {
            return 'downloaded_file';
        }
    }

    /**
     * Descarga archivos de múltiples páginas
     * @param {Array<string>} pageUrls - Array de URLs de páginas
     * @param {string} fileExtension - Extensión de archivos a buscar
     * @param {string} downloadDir - Directorio donde guardar los archivos
     * @returns {Promise<number>} Total de archivos descargados
     */
    async downloadFromMultiplePages(pageUrls, fileExtension, downloadDir = 'archivos_js') {
        let totalDownloads = 0;

        for (const pageUrl of pageUrls) {
            console.log(`\nProcesando página: ${pageUrl}`);
            const downloads = await this.downloadFilesFromPage(pageUrl, fileExtension, downloadDir);
            totalDownloads += downloads;
        }

        return totalDownloads;
    }
}

module.exports = FileDownloader;

File: src/java/FileDownloader.java
package com.filedownloader;

import java.io.*;
import java.net.*;
import java.nio.file.*;
import java.util.*;
import java.util.regex.*;
import java.util.concurrent.*;

/**
 * Clase principal para descargar archivos desde páginas web
 */
public class FileDownloader {
    private final String baseUrl;
    private final String userAgent;
    private final int timeout;
    private final ExecutorService executorService;

    /**
     * Constructor
     * @param baseUrl URL base del sitio web
     */
    public FileDownloader(String baseUrl) {
        this(baseUrl, "Mozilla/5.0 (Java FileDownloader)", 30000);
    }

    /**
     * Constructor con parámetros personalizados
     * @param baseUrl URL base del sitio web
     * @param userAgent User agent para las peticiones HTTP
     * @param timeout Timeout en milisegundos
     */
    public FileDownloader(String baseUrl, String userAgent, int timeout) {
        this.baseUrl = baseUrl;
        this.userAgent = userAgent;
        this.timeout = timeout;
        this.executorService = Executors.newFixedThreadPool(5);
    }

    /**
     * Obtiene el contenido HTML de una página
     * @param url URL de la página
     * @return Contenido HTML como String
     * @throws IOException Si hay error en la conexión
     */
    public String fetchPageContent(String url) throws IOException {
        HttpURLConnection connection = null;
        try {
            URL urlObj = new URL(url);
            connection = (HttpURLConnection) urlObj.openConnection();
            
            // Configurar conexión
            connection.setRequestMethod("GET");
            connection.setRequestProperty("User-Agent", userAgent);
            connection.setConnectTimeout(timeout);
            connection.setReadTimeout(timeout);
            
            // Verificar código de respuesta
            int responseCode = connection.getResponseCode();
            if (responseCode != HttpURLConnection.HTTP_OK) {
                throw new IOException("HTTP Error: " + responseCode);
            }
            
            // Leer contenido
            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(connection.getInputStream()))) {
                StringBuilder content = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    content.append(line).append("\n");
                }
                return content.toString();
            }
            
        } finally {
            if (connection != null) {
                connection.disconnect();
            }
        }
    }

    /**
     * Extrae enlaces de archivos con una extensión específica
     * @param htmlContent Contenido HTML
     * @param fileExtension Extensión de archivo a buscar
     * @return Lista de URLs de archivos
     */
    public List<String> parseFileLinks(String htmlContent, String fileExtension) {
        List<String> fileLinks = new ArrayList<>();
        
        // Expresión regular para encontrar enlaces href
        String regex = "href\\s*=\\s*['\"]\\s*([^'\"]*" + Pattern.quote(fileExtension) + "[^'\"]*)\\s*['\"]";
        Pattern pattern = Pattern.compile(regex, Pattern.CASE_INSENSITIVE);
        Matcher matcher = pattern.matcher(htmlContent);
        
        while (matcher.find()) {
            String link = matcher.group(1);
            String absoluteUrl = resolveUrl(baseUrl, link);
            fileLinks.add(absoluteUrl);
        }
        
        return fileLinks;
    }

    /**
     * Descarga un archivo desde una URL
     * @param fileUrl URL del archivo
     * @param outputPath Ruta donde guardar el archivo
     * @return true si la descarga fue exitosa
     */
    public boolean downloadFile(String fileUrl, String outputPath) {
        HttpURLConnection connection = null;
        try {
            // Crear directorio si no existe
            Path filePath = Paths.get(outputPath);
            Path parentDir = filePath.getParent();
            if (parentDir != null && !Files.exists(parentDir)) {
                Files.createDirectories(parentDir);
            }
            
            // Establecer conexión
            URL url = new URL(fileUrl);
            connection = (HttpURLConnection) url.openConnection();
            connection.setRequestProperty("User-Agent", userAgent);
            connection.setConnectTimeout(timeout);
            connection.setReadTimeout(timeout);
            
            // Verificar código de respuesta
            int responseCode = connection.getResponseCode();
            if (responseCode != HttpURLConnection.HTTP_OK) {
                System.err.println("Error HTTP " + responseCode + " descargando " + fileUrl);
                return false;
            }
            
            // Descargar archivo
            try (InputStream inputStream = connection.getInputStream();
                 FileOutputStream outputStream = new FileOutputStream(outputPath)) {
                
                byte[] buffer = new byte[8192];
                int bytesRead;
                while ((bytesRead = inputStream.read(buffer)) != -1) {
                    outputStream.write(buffer, 0, bytesRead);
                }
            }
            
            System.out.println("Descargado: " + extractFilename(fileUrl));
            return true;
            
        } catch (IOException e) {
            System.err.println("Error descargando " + fileUrl + ": " + e.getMessage());
            return false;
        } finally {
            if (connection != null) {
                connection.disconnect();
            }
        }
    }

    /**
     * Descarga todos los archivos con una extensión específica desde una página
     * @param pageUrl URL de la página
     * @param fileExtension Extensión de archivos a buscar
     * @param downloadDir Directorio donde guardar los archivos
     * @return Número de archivos descargados exitosamente
     */
    public int downloadFilesFromPage(String pageUrl, String fileExtension, String downloadDir) {
        try {
            // Obtener contenido de la página
            String htmlContent = fetchPageContent(pageUrl);
            
            // Parsear enlaces de archivos
            List<String> fileLinks = parseFileLinks(htmlContent, fileExtension);
            
            if (fileLinks.isEmpty()) {
                System.out.println("No se encontraron archivos con extensión '" + fileExtension + "'");
                return 0;
            }
            
            System.out.println("Encontrados " + fileLinks.size() + " archivos para descargar");
            
            // Crear directorio de descarga
            Path downloadPath = Paths.get(downloadDir);
            if (!Files.exists(downloadPath)) {
                Files.createDirectories(downloadPath);
            }
            
            // Descargar archivos en paralelo
            List<CompletableFuture<Boolean>> futures = new ArrayList<>();
            
            for (String fileUrl : fileLinks) {
                String filename = extractFilename(fileUrl);
                String outputPath = Paths.get(downloadDir, filename).toString();
                
                CompletableFuture<Boolean> future = CompletableFuture.supplyAsync(
                    () -> downloadFile(fileUrl, outputPath), executorService);
                futures.add(future);
            }
            
            // Esperar a que terminen todas las descargas
            int successfulDownloads = 0;
            for (CompletableFuture<Boolean> future : futures) {
                try {
                    if (future.get()) {
                        successfulDownloads++;
                    }
                } catch (InterruptedException | ExecutionException e) {
                    System.err.println("Error en descarga paralela: " + e.getMessage());
                }
            }
            
            return successfulDownloads;
            
        } catch (IOException e) {
            System.err.println("Error obteniendo la página: " + e.getMessage());
            return 0;
        }
    }

    /**
     * Convierte URL relativa a absoluta
     * @param baseUrl URL base
     * @param relativeUrl URL relativa
     * @return URL absoluta
     */
    public static String resolveUrl(String baseUrl, String relativeUrl) {
        try {
            URI baseUri = new URI(baseUrl);
            URI resolvedUri = baseUri.resolve(relativeUrl);
            return resolvedUri.toString();
        } catch (URISyntaxException e) {
            System.err.println("Error resolviendo URL: " + e.getMessage());
            return relativeUrl;
        }
    }

    /**
     * Extrae el nombre del archivo de una URL
     * @param url URL del archivo
     * @return Nombre del archivo
     */
    public static String extractFilename(String url) {
        try {
            URI uri = new URI(url);
            String path = uri.getPath();
            if (path != null && !path.isEmpty()) {
                String[] parts = path.split("/");
                if (parts.length > 0) {
                    String filename = parts[parts.length - 1];
                    return filename.isEmpty() ? "downloaded_file" : filename;
                }
            }
        } catch (URISyntaxException e) {
            System.err.println("Error extrayendo nombre de archivo: " + e.getMessage());
        }
        return "downloaded_file";
    }

    /**
     * Cierra el executor service
     */
    public void shutdown() {
        executorService.shutdown();
        try {
            if (!executorService.awaitTermination(60, TimeUnit.SECONDS)) {
                executorService.shutdownNow();
            }
        } catch (InterruptedException e) {
            executorService.shutdownNow();
            Thread.currentThread().interrupt();
        }
    }
}